copycat/api/src/main/java/org/apache/kafka/copycat/data/SchemaProjector.java

```diff
@@ -0,0 +1,209 @@
+        }
+        Type writerValueType = writer.valueSchema().type();
+        Type readerValueType = reader.valueSchema().type();
+        if (writerValueType != readerValueType && !promote(writerValueType, readerValueType)) {
+            throw new SchemaProjectorException("Value schema type mismatch for array. Writer value schema type: " + writer.type() + " and Reader value schema type: " + reader.type());
+        }
+        List<?> array = (List<?>) record;
+        List<Object> retArray = new ArrayList<>();
+        for (Object entry : array) {
+            retArray.add(project(writer.valueSchema(), reader.valueSchema(), entry));
+        }
+        return retArray;
+    }
+
+    private static Object projectMap(Schema writer, Schema reader, Object record) throws SchemaProjectorException {
+        if (writer.type() != reader.type()) {
+            throw new SchemaProjectorException("Schema type mismatch. Writer type: " + writer.type() + " and Reader type: " + reader.type());
+        }
+        Type writerKeyType = writer.keySchema().type();
+        Type readerKeyType = reader.keySchema().type();
+        if (writerKeyType != readerKeyType && !promote(writerKeyType, readerKeyType)) {
+            throw new SchemaProjectorException("Key schema type mismatch for array. Writer key schema type: " + writer.type() + " and Reader key schema type: " + reader.type());
+        }
+        Type writerValueType = writer.valueSchema().type();
+        Type readerValueType = reader.valueSchema().type();
+        if (writerValueType != readerValueType && !promote(writerValueType, readerValueType)) {
+            throw new SchemaProjectorException("Value schema type mismatch for array. Writer value schema type: " + writer.type() + " and Reader value schema type: " + reader.type());
+        }
+        Map<?, ?> map = (Map<?, ?>) record;
+        Map<Object, Object> retMap = new HashMap<>();
+        for (Map.Entry<?, ?> entry : map.entrySet()) {
+            Object key = entry.getKey();
+            Object value = entry.getValue();
+            Object retKey = project(writer.keySchema(), reader.keySchema(), key);
+            Object retValue = project(writer.valueSchema(), writer.valueSchema(), value);
+            retMap.put(retKey, retValue);
+        }
+        return retMap;
+    }
+
+    private static Object projectLogical(Schema writer, Schema reader, Object record) throws SchemaProjectorException {
+        if (writer.name() == null || !writer.name().equals(reader.name())) {
+            throw new SchemaProjectorException("Schema does not match for " + writer.name() + " and " + reader.name());
+        }
+
+        switch (reader.name()) {
+            case Decimal.LOGICAL_NAME:
+            case Date.LOGICAL_NAME:
+            case Time.LOGICAL_NAME:
+            case Timestamp.LOGICAL_NAME:
+                return record;
+            default:
+                throw new SchemaProjectorException("Not a valid logical type: " + reader.name());
```

Comment:
I think we just want to get rid of the special handling for logical types, but if it stays, the handling here needs to change. Any user defined schema can have a name too.